### PR TITLE
FIX-#4564: Register custom serializer to avoid Ray race condition

### LINKED ISSUE #4564: BUG: ray 1.13 breaks CI, including test_binary and test_pickle, with circular import
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macOS Monterey 12.2.1
- **Modin version** (`modin.__version__`): ba2b6ea47d81575e41af6c92ba4c2c3f8bcf54e2
- **Python version**: 3.9.12
- **Code we can use to reproduce**:

With ray 1.13, run

`pytest modin/pandas/test/dataframe/test_pickle.py`

and

`pytest modin/pandas/test/dataframe/test_binary.py` 

and maybe more.

Sample CI failure: https://github.com/modin-project/modin/runs/6819134820?check_suite_focus=true

### Describe the problem
I can't tell what's going wrong. The ray tasks raise a `RaySystemError` with stack trace for ` modin/pandas/test/dataframe/test_pickle.py::test_column_pickle[True]`:

<details>
<summary>Stack trace</summary>

```
ray.exceptions.RaySystemError: System error: cannot import name 'DataFrame' from partially initialized module 'pandas.core.frame' (most likely due to a circular import) (/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/frame.py)
traceback: Traceback (most recent call last):
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/ray/serialization.py", line 340, in deserialize_objects
    obj = self._deserialize_object(data, metadata, object_ref)
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/ray/serialization.py", line 237, in _deserialize_object
    return self._deserialize_msgpack_data(data, metadata_fields)
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/ray/serialization.py", line 192, in _deserialize_msgpack_data
    python_objects = self._deserialize_pickle5_data(pickle5_data)
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/ray/serialization.py", line 180, in _deserialize_pickle5_data
    obj = pickle.loads(in_band, buffers=buffers)
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/frame.py", line 157, in <module>
    from pandas.core.generic import NDFrame
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/generic.py", line 114, in <module>
    from pandas.core import (
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/indexing.py", line 45, in <module>
    import pandas.core.common as com
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/__init__.py", line 48, in <module>
    from pandas.core.api import (
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/api.py", line 48, in <module>
    from pandas.core.groupby import (
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/groupby/__init__.py", line 1, in <module>
    from pandas.core.groupby.generic import (
  File "/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/groupby/generic.py", line 70, in <module>
    from pandas.core.frame import DataFrame
ImportError: cannot import name 'DataFrame' from partially initialized module 'pandas.core.frame' (most likely due to a circular import) (/Users/maheshvashishtha/opt/anaconda3/envs/modin-latest/lib/python3.9/site-packages/pandas/core/frame.py)
```
</details>

For `modin/pandas/test/dataframe/test_binary.py` for some reason I can't reproduce the failure on my laptop but here it is in CI: https://github.com/modin-project/modin/runs/6819134820?check_suite_focus=true

### PR DESCRIPTION
**Pull request by devin-petersohn**
Signed-off-by: Devin Petersohn <devin.petersohn@gmail.com>

## What do these changes do?

- [ ] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [ ] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] signed commit with `git commit -s` 
- [ ] Resolves #4564  
- [ ] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date 
- [ ] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

**Review by RehanSD — DISMISSED** [review]
LGTM! Pleased to release notes!

File: modin/core/execution/ray/common/utils.py [hunk 1]
```diff
@@ -224,6 +226,9 @@ def initialize_ray(
         _move_stdlib_ahead_of_site_packages
     )
     ray.worker.global_worker.run_function_on_all_workers(_import_pandas)
+    ray.util.register_serializer(
```
Comment (YarShev) [inline modin/core/execution/ray/common/utils.py:229]: I think we should add a comment on why we do this.

**Review by YarShev — COMMENTED** [review]
Please fix the PR title, link an issue to be resolved and add a release note with your GH nickname.

**Comment by mvashishtha** [comment]
@devin-petersohn You have "customer" instead of "custom" in the PR title and first commit description.

**Review by mvashishtha — APPROVED** [review]
LGTM, Thanks!

**Review by RehanSD — APPROVED** [review]
LGTM! We should conduct a perf check to see how `pickle` compares to Ray's serializer.

**Comment by devin-petersohn** [comment]
>@devin-petersohn You have "customer" instead of "custom" in the PR title and first commit description.

:rofl: fixed

**Comment by mvashishtha** [comment]
@devin-petersohn "customer" is still in the first commit description. It's fine by me to leave it that way.

**Review by RehanSD — CHANGES_REQUESTED** [review]
Please add your name to the contributors in the release notes!

File: docs/release_notes/release_notes-0.16.0.rst [hunk 2]
```diff
@@ -27,7 +27,7 @@ Key Features and Updates
 * Documentation improvements
   * DOCS-#4552: Change default sphinx language to en to fix sphinx >= 5.0.0 build (#4553)
 * Dependencies
-  *
+  * FIX-#4564: Register customer serializer to avoid Ray race condition (#4568)
```
Comment (YarShev) [inline docs/release_notes/release_notes-0.16.0.rst:30]:
```suggestion
  * FIX-#4564: Register custom serializer to avoid Ray race condition (#4568)
```

**Review by naren-ponder — APPROVED** [review]

**Review by RehanSD — APPROVED** [review]
LGTM!

**Review by YarShev — APPROVED** [review]

**Comment by devin-petersohn** [comment]
@modin-project/modin-ray is this the recommended way of fixing this? It seems like a bad idea but it is the only way we could use the latest ray release without a race condition

**Review by vnlitvinov — APPROVED** [review]
LGTM, though I would like to find a better way... I wish it existed 😞

**Comment by vnlitvinov** [comment]
@devin-petersohn a good point here:

> LGTM! We should conduct a perf check to see how pickle compares to Ray's serializer.

Could someone please check some benchmarks?

**Comment by NickCrews** [comment]
Drive-by-ing here, IDK if you have a good way of running benchmarks, but perhaps the [github-actions bot](https://github.com/dedupeio/dedupe/blob/main/.github/workflows/benchmark-bot.yml) that we wrote for the dedupe library could be useful to you. Take a look at [this comment](https://github.com/dedupeio/dedupe/pull/1062#issuecomment-1152954350) to see how it is used. Makes it really easy for us to test performance changes on a PR.

**Comment by devin-petersohn** [comment]
Closing because we chose to go a different direction.

**Pull request closed without merging**